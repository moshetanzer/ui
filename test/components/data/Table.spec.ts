import { describe, it, expect } from 'vitest'
import { UTable } from '#components'
import type { TypeOf } from 'zod'
import ComponentRender from '../component-render'

describe('Table', () => {
  it.each([
    [
      'basic case',
      {
        props: {
          columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'sort',
      {
        props: {
          sort: { column: 'name', direction: 'desc' }, columns: [{ key: 'name', label: 'Name', sortable: true }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'sort mode',
      {
        props: {
          sortMode: 'manual', columns: [{ key: 'name', label: 'Name', sortable: true }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'custom sort button',
      {
        props:
        {
          sortButton: { label: 'Custom button', size: 'xl', variant: 'orange', icon: 'i-heroicons-academic-cap' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'sort asc icon',
      {
        props:
        {
          sortAscIcon: 'i-heroicons-adjustments-horizontal', sort: { column: 'name', direction: 'asc' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'sort desc icon',
      {
        props: {
          sortDescIcon: 'i-heroicons-arrow-down-left-solid', sort: { column: 'name', direction: 'desc' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'loading state',
      {
        props:
        {
          loading: true, loadingState: { icon: 'i-heroicons-sparkles', label: 'Loading...' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    [
      'loading state slot',
      {
        slots: { 'loading-state': () => 'Loading state slot' },
        props:
        {
          loading: true, loadingState: { icon: 'i-heroicons-sparkles', label: 'Loading...' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    ['empty state props',
      {
        props:
        {
          rows: [], emptyState: { icon: 'i-heroicons-inbox', label: 'No data available' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    ['empty state slot',
      {
        slots: { 'empty-state': () => 'Empty state slot' },
        props: {
          emptyState: { label: 'emptyStateLabel', icon: 'i-heroicons-chevron-up-down-solid' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ],
    ['selected rows',
      {
        props: {
          rows: [{ id: 1, name: 'John' }, { id: 2, name: 'Jane' }, { id: 3, name: 'Doe' }], columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }]
        }
      }
    ]

    // @ts-ignore
  ])('renders %s correctly', async (nameOrHtml: string, options: TypeOf<typeof UTable.props>) => {
    const html = await ComponentRender(nameOrHtml, options, UTable)
    expect(html).toMatchSnapshot()
  })
})
