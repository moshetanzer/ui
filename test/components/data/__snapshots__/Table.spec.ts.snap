// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders basic case correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders custom sort button correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-base gap-x-2.5 p-2.5 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-academic-cap flex-shrink-0 h-6 w-6" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders empty state props correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-inbox w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No data available</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders empty state slot correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">Empty state slot</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders loading state correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <tr>
        <td colspan="0" class="absolute inset-x-0 -bottom-[0.5px] p-0">
          <div data-v-0f229b8a="" class="w-full flex flex-col gap-2">
            <!--v-if--><progress data-v-0f229b8a="" class="block appearance-none border-none overflow-hidden w-full [&::-webkit-progress-bar]:w-full h-px rounded-full [&::-webkit-progress-bar]:rounded-full [&::-webkit-progress-bar]:bg-gray-200 [&::-webkit-progress-bar]:dark:bg-gray-700 [@supports(selector(&::-moz-progress-bar))]:bg-gray-200 [@supports(selector(&::-moz-progress-bar))]:dark:bg-gray-700 [&::-webkit-progress-value]:rounded-full [&::-webkit-progress-value]:transition-all [&::-webkit-progress-value]:ease-in-out [&::-moz-progress-bar]:rounded-full text-primary-500 dark:text-primary-400 [&::-webkit-progress-value]:bg-current [&::-moz-progress-bar]:bg-current indeterminate:relative indeterminate:after:rounded-full [&:indeterminate::-webkit-progress-value]:rounded-full [&:indeterminate::-moz-progress-bar]:rounded-full bar-animation-carousel"></progress>
            <!--v-if-->
          </div>
        </td>
      </tr>
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-sparkles w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4 animate-spin" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">Loading...</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders loading state slot correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <tr>
        <td colspan="0" class="absolute inset-x-0 -bottom-[0.5px] p-0">
          <div data-v-0f229b8a="" class="w-full flex flex-col gap-2">
            <!--v-if--><progress data-v-0f229b8a="" class="block appearance-none border-none overflow-hidden w-full [&::-webkit-progress-bar]:w-full h-px rounded-full [&::-webkit-progress-bar]:rounded-full [&::-webkit-progress-bar]:bg-gray-200 [&::-webkit-progress-bar]:dark:bg-gray-700 [@supports(selector(&::-moz-progress-bar))]:bg-gray-200 [@supports(selector(&::-moz-progress-bar))]:dark:bg-gray-700 [&::-webkit-progress-value]:rounded-full [&::-webkit-progress-value]:transition-all [&::-webkit-progress-value]:ease-in-out [&::-moz-progress-bar]:rounded-full text-primary-500 dark:text-primary-400 [&::-webkit-progress-value]:bg-current [&::-moz-progress-bar]:bg-current indeterminate:relative indeterminate:after:rounded-full [&:indeterminate::-webkit-progress-value]:rounded-full [&:indeterminate::-moz-progress-bar]:rounded-full bar-animation-carousel"></progress>
            <!--v-if-->
          </div>
        </td>
      </tr>
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">Loading state slot</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders selected rows correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Name</span></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr class="">
        <!--v-if-->
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm">John</td>
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"></td>
      </tr>
      <tr class="">
        <!--v-if-->
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm">Jane</td>
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"></td>
      </tr>
      <tr class="">
        <!--v-if-->
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm">Doe</td>
        <td class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"></td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort asc icon correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort desc icon correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort mode correctly 1`] = `
"<div class="relative overflow-x-auto">
  <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
    <thead class="relative">
      <tr class="">
        <!--v-if-->
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>Age</span></th>
      </tr>
      <!--v-if-->
    </thead>
    <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4" aria-hidden="true"></span>
            <p class="text-sm text-center text-gray-900 dark:text-white">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;
